base/ca/src/com/netscape/ca/CAService.java

```diff
@@ -896,25 +899,70 @@ private X509CertImpl issueX509Cert(
+                        exts.delete("1.3.6.1.4.1.11129.2.4.3");
+                        // debug print without poison ext
+                        // printExtensions(exts);
+                        certi.delete(X509CertInfo.EXTENSIONS);
+                        certi.set(X509CertInfo.EXTENSIONS, exts);
+
+                        // tbsCert_save_s is the tbsCert without poison ext;
+                        // Save for debugging later
+                        byte[] tbsCert_save = certi.getEncodedInfo(true);
+                        String tbsCert_save_s = 
+                            CryptoUtil.base64Encode(tbsCert_save);
+                        //logger.debug(method + "tbsCert without poison="
+                        //        + tbsCert_save_s);
+
+                        // now add back
+                        exts.set("1.3.6.1.4.1.11129.2.4.3", ctPoison);
+                        //printExtensions(exts);
+                        certi.delete(X509CertInfo.EXTENSIONS);
+                        certi.set(X509CertInfo.EXTENSIONS, exts);
+
                         logger.debug(method + " About to ca.sign CT pre-cert.");
                         cert = ca.sign(certi, algname);
 
                         /*
                          * remove the poison extension from certi
                          */
                         exts.delete("1.3.6.1.4.1.11129.2.4.3");
+                        logger.debug(method + " ctPoison deleted");
+                        // debug print without poison ext
+                        // printExtensions(exts);
                         certi.delete(X509CertInfo.EXTENSIONS);
                         certi.set(X509CertInfo.EXTENSIONS, exts);
                         /*
-                         * This tbsCert does not contain poison ext.
-                         * It is intended to be used for verifying
-                         * the SCT signature in the CT response,
+                         * tbsCert is the tbsCert after
+                         *   - poision ext was deleted,
+                         *   - then added back,
+                         *   - then deleted again
+                         * It can be used for comparison with
+                         * tbsCert_save_s to check if things get
+                         * out of order for debugging purpose.
+                         *
+                         * It is also intended to be used for verifying
+                         * the SCT signature in the CT response later,
                          * if possible (see verifySCT for detail)
                          */
-                        byte[] tbsCert = certi.getEncodedInfo();
+                        byte[] tbsCert = certi.getEncodedInfo(true);
+
+                        // debugging: test to see if things got out of order
+                        String tbsCert_s = CryptoUtil.base64Encode(tbsCert);
+                        if (tbsCert_save_s.equals(tbsCert_s)) {
```

Comment:
I am not clear why we are comparing here? Are you trying to ensure the order of extensions remains the same? There is no guarantee the order will be preserved, at least until the changes are made in JSS. Then this comparison won't be required.

Reply:
Yes, that's correct.  As I explained earlier, that is the case.  However, in my experiments so far when doing printExtensions after delete/add/delete, the "order" did stay.  Again, this patch is simply to get to Fraser quickly so he could help debug the verifySCT sooner, so please ignore anything that looks hacky for now. Sorry I should have explained more clearly. ;-)